Updates measurements page (#372)
XanaduAI/pennylane: doc/introduction/measurements.rst

.. currentmodule:: pennylane.measure

PennyLane can extract different types of measurement results: the expectation of an observable
over multiple measurements, its variance, or samples of a single measurement.

> For example, the quantum function shown in the previous section

Reviewer: I think it's better if we minimize these sequential references and make every page standalone (even if content is repeated). Users could land on this page as the first place they click. (also makes it more maintainable if the other page later changes)

Author: Agree, will change.

used the :func:`~pennylane.expval` measurement:

.. code-block:: python

    def my_quantum_function(x, y):
        qml.RZ(x, wires=0)
        qml.CNOT(wires=[0, 1])
        qml.RY(y, wires=1)
        return qml.expval(qml.PauliZ(1))

> For more details on the available measurement functions:

Reviewer: This feels a bit weird. First off, there's no verb (i.e., check out the following links). But more importantly, there isn't really any content on those pages (just one-liner docstrings). Suggest rewriting this part

Author: Changed to:
> The available measurement functions are:
Future-proofs us in case we add more (i.e., covariance) :)

.. currentmodule:: pennylane.measure

PennyLane can extract different types of measurement results from quantum
devices: the expectation of an observable, its variance, or
samples of a single measurement.

> For example, a QNode that returns the expectation value of the

Reviewer: ```suggestion
For example, the following circuit returns the expectation value of the
```

:html:`</div>`

.. note::

    Both :func:`~.pennylane.expval` and :func:`~.pennylane.var` support analytic
    differentiation. :func:`~.pennylane.sample`, however, returns *stochastic*
    results, and does not support differentiation.

Multiple measurements
---------------------

Quantum functions can also return multiple measurements, as long as each wire
is not measured more than once:

.. code-block:: python

    def my_quantum_function(x, y):
        qml.RZ(x, wires=0)
        qml.CNOT(wires=[0, 1])
        qml.RY(y, wires=1)
        return qml.expval(qml.PauliZ(1)), qml.var(qml.PauliX(0))

You can also use list comprehensions, and other common Python patterns:

.. code-block:: python

    def my_quantum_function(x, y):
        qml.RZ(x, wires=0)
        qml.CNOT(wires=[0, 1])
        qml.RY(y, wires=1)
        return [qml.expval(qml.PauliZ(i)) for i in range(2)]

Tensor observables
------------------

PennyLane supports measuring the tensor product of observables, by using

> the ``@`` notation. For example, to measure the expectation value of

Reviewer: do we want to mention that the `Tensor` class can also be used?

Author: I thought that might be confusing/overcomplicate - thoughts?

Reviewer: fair enough

:math:`Z\otimes I \otimes X`:

.. code-block:: python3

    def my_quantum_function(x, y):
        qml.RZ(x, wires=0)
        qml.CNOT(wires=[0, 1])
        qml.RY(y, wires=1)
        qml.CNOT(wires=[0, 2])

>         return qml.expval(qml.PauliZ(0) @ qml.PauliX(2))

Reviewer: maybe be good to note that there's no need to declare the identity term?

Author: I thought that might be confusing/overcomplicate - thoughts?

Author: done

    results, and does not support differentiation.

Multiple measurements
---------------------

Quantum functions can also return multiple measurements, as long as each wire
is not measured more than once:

.. code-block:: python

    def my_quantum_function(x, y):
        qml.RZ(x, wires=0)
        qml.CNOT(wires=[0, 1])
        qml.RY(y, wires=1)
        return qml.expval(qml.PauliZ(1)), qml.var(qml.PauliX(0))

You can also use list comprehensions, and other common Python patterns:

.. code-block:: python

    def my_quantum_function(x, y):
        qml.RZ(x, wires=0)
        qml.CNOT(wires=[0, 1])
        qml.RY(y, wires=1)
        return [qml.expval(qml.PauliZ(i)) for i in range(2)]

Tensor observables
------------------

PennyLane supports measuring the tensor product of observables, by using
the ``@`` notation. For example, to measure the expectation value of
:math:`Z\otimes I \otimes X`:

.. code-block:: python3

    def my_quantum_function(x, y):
        qml.RZ(x, wires=0)
        qml.CNOT(wires=[0, 1])
        qml.RY(y, wires=1)
        qml.CNOT(wires=[0, 2])
        return qml.expval(qml.PauliZ(0) @ qml.PauliX(2))

> Note that we don't need to declare the identity term on wire 1; this is

Reviewer: ```suggestion
Note that we don't need to declare the identity observable on wire 1; this is
```
